app/modules/uart.c

```diff
@@ -7,91 +7,65 @@
 #include "c_types.h"
 #include "c_string.h"
 #include "rom.h"
+#include "driver/input.h"
 
 static int uart_receive_rf = LUA_NOREF;
-bool run_input = true;
-bool uart_on_data_cb(const char *buf, size_t len){
-  if(!buf || len==0)
-    return false;
-  if(uart_receive_rf == LUA_NOREF)
-    return false;
+
+void uart_on_data_cb(const char *buf, size_t len){
   lua_State *L = lua_getstate();
-  if(!L)
-    return false;
   lua_rawgeti(L, LUA_REGISTRYINDEX, uart_receive_rf);
   lua_pushlstring(L, buf, len);
-  lua_call(L, 1, 0);
-  return !run_input;
+  luaN_call(L, 1, 0, 0);
```

Comment:
So here for example, I believe if the call fails we've just littered the stack with an error message.

Reply:
> The real error is [here](https://travis-ci.org/nodemcu/nodemcu-firmware/jobs/560814634#L3867) in the build log:

Yup, I'd already noticed that one. :cry:

Reply:
This actually touches on a more systemic issue in that all CB action routines should always balance the stack from entry to return but we don't have any guidance / quality check / coding pattern to ensure this and failure to do so leads to resource leakage that will ultimately exhaust memory.   We need to address this wider issue, IMO.

However as to this specific point, we need to decide the stack rules for `luaN_call()`.  IMO we have two options:
1.  It follows the `lua_pcall()`convention of returning 1 argument and an error status if the called routine throws an error, or
2.  It dummies the `lua_call()` of returning `nresults` arguments (which might be nil in the case of an error) and an error status.

IIRC, there only one use in the modules library where the library routine accepts _any_ result from the Lua CB, and all that typically happens is that the wrapper function returns control to the SDK on return from the CB.  I think that we should go for easy of migration to `luaN_call()` and that adopting option 2 will best achieve this.

Reply:
My vote would be for #2 as well. If there are any instances where a module is actually interested in the error message, they can use `lua_pcall()` for that. For everyone else, I think the simplicity of `lua_call()` behaviour will give us the most robust outcome.